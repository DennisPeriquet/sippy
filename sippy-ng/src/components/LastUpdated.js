import PropTypes from 'prop-types'
import React, { Fragment } from 'react'

export default function LastUpdated (props) {
  const minute = 1000 * 60 // Milliseconds in a minute
  const hour = 60 * minute // Milliseconds in an hour
  const millisAgo = props.lastUpdated ? (props.lastUpdated.getTime() - Date.now()) : 0

  try {
    if (millisAgo === 0) {
      return <Fragment>Last update unknown</Fragment>
    } else if (Math.abs(millisAgo) < hour) {
      return (
                <Fragment>
                    Last updated {Math.round(Math.abs(millisAgo) / minute) + ' minutes ago'}
                </Fragment>
      )
    } else {
      return (
                <Fragment>
                    Last updated {Math.round(Math.abs(millisAgo) / hour) + ' hours ago'}
                </Fragment>
      )
    }
  } catch (e) {
    return <></>
  }
}

LastUpdated.propTypes = {
  lastUpdated: PropTypes.instanceOf(Date).isRequired
}
