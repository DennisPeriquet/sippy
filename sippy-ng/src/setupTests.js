/** @jest-environment setup-polly-jest/jest-environment-node */

import '@testing-library/jest-dom'
import Enzyme from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import fetch from 'node-fetch'
import path from 'path'
import { setupPolly } from 'setup-polly-jest'
import toDiffableHtml from 'diffable-html'

// https://github.com/mui-org/material-ui/issues/21293
export const withoutMuiID = (wrapper) => toDiffableHtml(wrapper.html()
  .replace(/id="mui-[0-9]*"/g, '')
  .replace(/aria-labelledby="(mui-[0-9]* *)*"/g, ''))

Enzyme.configure({ adapter: new Adapter() })

// To stub out fetch(), we need to use node-fetch and this line:
global.fetch = fetch

// See: https://github.com/vuejs/vue-test-utils/issues/974
global.requestAnimationFrame = cb => cb()

// When injecting multiple filters, material table complains.
const originalError = console.error.bind(console.error)

global.console.error = (log) => {
  if (!log.toString().includes('Warning: Failed')) {
    originalError(log.toString())
  }
}

// Set API URL for Sippy
process.env.REACT_APP_API_URL = 'http://localhost:8080'

// Default PollyJS mode is replay, set to record to record
// new API calls.
if (!process.env.POLLY_MODE) {
  process.env.POLLY_MODE = 'replay'
}

export const setupDefaultPolly = () => {
  const context = setupPolly({
    mode: process.env.POLLY_MODE,
    adapters: [require('@pollyjs/adapter-node-http')],
    persister: require('@pollyjs/persister-fs'),
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, '../__recordings__')
      }
    }
  })
  return context
}

export const expectLoadingPage = (wrapper) => expect(wrapper.find('p').contains('Loading...'))

jest.mock('react-chartjs-2', () => ({
  Bar: () => null,
  Chart: () => null,
  ChartComponent: () => null,
  Doughnut: () => null,
  Line: () => null,
  PolarArea: () => null
}))
