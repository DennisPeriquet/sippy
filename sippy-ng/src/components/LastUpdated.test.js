import React from 'react'
import { act } from '@testing-library/react'
import { withoutMuiID } from '../setupTests'
import LastUpdated from './LastUpdated'
import { mount } from 'enzyme'

const minute = 60 * 1000
const hour = minute * 60

jest.useRealTimers()

describe(LastUpdated, () => {
  it('shows unknown when last update is unknown', async () => {
    let wrapper
    Date.now = jest.fn(() => 0)

    await act(async () => {
      wrapper = mount(<LastUpdated lastUpdated={new Date(0)} />)
    })

    expect(wrapper.html()).toContain('Last update unknown')
    expect(withoutMuiID(wrapper)).toMatchSnapshot()
  })

  it('shows minutes when < an hour', async () => {
    const lastUpdated = new Date(0)
    Date.now = jest.fn(() => 10 * minute)

    let wrapper
    await act(async () => {
      wrapper = mount(<LastUpdated lastUpdated={lastUpdated} />)
    })

    expect(wrapper.html()).toContain('Last updated 10 minutes ago')
    expect(withoutMuiID(wrapper)).toMatchSnapshot()
  })

  it('shows hours when > an hour', async () => {
    const lastUpdated = new Date(0)
    Date.now = jest.fn(() => 3 * hour)

    let wrapper
    await act(async () => {
      wrapper = mount(<LastUpdated lastUpdated={lastUpdated} />)
    })

    expect(wrapper.exists()).toBe(true)
    expect(wrapper.html()).toContain('Last updated 3 hours ago')
    expect(withoutMuiID(wrapper)).toMatchSnapshot()
  })
})
