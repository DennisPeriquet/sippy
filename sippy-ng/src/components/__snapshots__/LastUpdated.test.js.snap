// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastUpdated shows hours when > an hour 1`] = `
"
Last updated 3 hours ago
"
`;

exports[`LastUpdated shows minutes when < an hour 1`] = `
"
Last updated 10 minutes ago
"
`;

exports[`LastUpdated shows unknown when last update is unknown 1`] = `
"
Last update unknown
"
`;
